Week3/HW03.hs

```diff
@@ -0,0 +1,169 @@
+module HW03 where
+
+-- https://www.cis.upenn.edu/~cis194/spring15/hw/03-ADTs.pdf
+
+import Data.Bool
+
+data Expression =
+    Var String                   -- Variable
+  | Val Int                      -- Integer literal
+  | Op Expression Bop Expression -- Operation
+  deriving (Show, Eq)
+
+-- Binary (2-input) operators
+data Bop = 
+    Plus     
+  | Minus    
+  | Times    
+  | Divide   
+  | Gt
+  | Ge       
+  | Lt  
+  | Le
+  | Eql
+  deriving (Show, Eq)
+
+data Statement =
+    Assign   String     Expression
+  | Incr     String
+  | If       Expression Statement  Statement
+  | While    Expression Statement       
+  | For      Statement  Expression Statement Statement
+  | Sequence Statement  Statement        
+  | Skip
+  deriving (Show, Eq)
+
+type State = String -> Int 
+
+-- Exercise 1 -----------------------------------------
+
+extend :: State -> String -> Int -> State
+extend initState name val a             = bool (initState a) val (a == name)
```

Comment:
<img width="1090" alt="Screenshot 2021-01-20 at 1 51 47 AM" src="https://user-images.githubusercontent.com/30658517/105088555-0f31a880-5ac2-11eb-8e6c-f600710ba9e8.png">

- i did it the first way that you suggested, but then the compiler suggested to move the 'name' outside so i did, but I shoudn't have listened, after all its's not the one who has to understand my code, I am hahah.

- Got it about the guards, will try use them more often than the bool of if-then-else which i keep using everywhere

Reply:
This is not compiler suggestion, just [hlint](https://github.com/ndmitchell/hlint) suggestion (haskell language server includes built-in support for it). But in this case the suggestion sounds reasonable. Even I'd do it because it's more concise :-)

Nevertheless the following two ways of defining a function are equivalent.
`f = \x -> x + 1`
`f x = x + 1`

Reply:
And maybe a bit more explanation about guards. Why I think they're often preferred over if-then-else:

1. you can have multiple guards per functions, which makes for more concise syntax than nested if-then-else:

```haskel
f x y | x > 0 = ...
        | x <0 && y < 5 = ...
        | otherwise = ...

-- versus
f x y = if x > 0 then ...
           else if x <0 && y < 5 then ...
           else ...
```
2. you can combine guards with pattern matching
```
data Pet = Cat Int | Dog String
f (Cat age)
    | age > 10 = ...
    | ...
f (Dog name)
    | length name < 5 = ...
    | otherwise = ...
```
3. you can use [pattern guards](https://wiki.haskell.org/Pattern_guard) which give you a lot more flexibility on top of that

Reply:
>  Ok, so it's resolver lts-16.28, so I can check https://www.stackage.org/lts-16.28 and see that it uses ghc-8.8.4


Looks like I have the same ghci version but a lower resolver:
<img width="896" alt="Screenshot 2021-01-20 at 8 08 18 PM" src="https://user-images.githubusercontent.com/30658517/105189699-3e482880-5b5b-11eb-8a0d-ab948463d7e6.png">